mechanical changes to eliminate insert## method

asraa opened the pull request:
* Adds inline header methods to modify header values
* Mechanical changes to `HeaderMap` API:
    - `headers.insertPath().value().setCopy(value)` becomes `headers.setCopyPath(value)`.
    - `headers.insertStatus().value().setInteger(value)` becomes `headers.setIntegerStatus(value)`.
    - `headers.insertContentType().value().setReference(value)` becomes `headers.setReferenceContentType(value)`
    - `headers.insertPath().value(value)` does an underlying copy, so this is transformed to `setCopy`.

The goal here is to remove accessing non-const `HeaderEntry` refs/ptrs. These mostly exist through the pattern,
`header_map.insertPath().value().setCopy(value)`,
where an inline header entry is accessed via the `insert##name()` macro and it's value is then modified. Removing the ability to grab modifiable `HeaderEntry`'s would allow `HeaderMap` to process all calls through it's API and keep track of it's internal byte size.

This PR introduces macros to set inline header values with the same performance as the above line. The use of this macro bypasses the need to do a lookup when using `HeaderMap` API calls, as in `header_map.setReferenceKey(Headers::get().Path, value)`.

The remaining non-const `HeaderEntry` accessors are because of appending or adding to values using the `HeaderString::apend` method or the static `appendToHeader` method. I will take care of those in a followup PR. After this, the non-const inline header accessors and the non-const `HeaderMap::get` method can be removed.

Part of https://github.com/envoyproxy/envoy/issues/8567
Testing: Added benchmarks to verify that the new calls perform just as fast as the previous calls.



Signed-off-by: Asra Ali <asraa@google.com>

jmarantz commented on include/envoy/http/header_map.h:
```diff
@@ -352,6 +352,9 @@ class HeaderEntry {
   virtual const HeaderEntry* name() const PURE;                                                    \
   virtual HeaderEntry* name() PURE;                                                                \
   virtual HeaderEntry& insert##name() PURE;                                                        \
```
do we want to remove insert in this PR? Or are there still callers we need to support?

Maybe this is worth a comment?

asraa replied:
Yep still need to support appending, edited the PR description.

jmarantz replied:
Should we replace this with an append() method then? Maybe you wouldn't need a non-const ref return at all then? Would that make it easier to reason about?

asraa replied:
I think that'd be the right approach. `insert` is probably the right name for this.

jmarantz commented on source/common/http/conn_manager_utility.cc:
```diff
@@ -387,7 +385,7 @@ void ConnectionManagerUtility::mutateResponseHeaders(HeaderMap& response_headers
         (!response_headers.TransferEncoding() && !response_headers.ContentLength());
     if (no_body) {
-      response_headers.insertContentLength().value(uint64_t(0));
+      response_headers.setIntegerContentLength(uint64_t(0));
```
WDYT of spelling this setContentLength() ? There would be no confusion with the string-ref or string_view versions which have specialized names. 

Actually you could also use `setForwardedClientCert(absl::string_view)`, and let the C++ type system figure out which one to pick.

Only the string-ref version would need to be named something different, which makes sense; there the caller is making an explicit promise that the ref will outlive the headers.

mattklein123 replied:
My concern here would be automatic conversions. IMO it's better to be explicit unless we are sure the auto conversions won't happen. Will defer to better C++ experts on this one.

mattklein123 commented on source/common/http/header_map_impl.cc:
```diff
@@ -616,6 +616,7 @@ HeaderMapImpl::HeaderEntryImpl& HeaderMapImpl::maybeCreateInline(HeaderEntryImpl
     return **entry;
   }
 
+  addSize(key.get().size());
```
Was this a previous bug?

asraa replied:
Ah. Nope, there was no `addSize` here because the method invalidated the cached size right above this.
I should remove that reset. It was there because before this PR the only caller of this method was `insert##name()` which also has resets byte size. Now there are `HeaderMapImpl` methods that are safe for tracking. I'll add in a test case into `header_map_impl_test` for this.

mattklein123 replied:
OK got it.

mattklein123 commented on source/common/http/header_map_impl.h:
```diff
@@ -33,6 +33,18 @@ public:
     return maybeCreateInline(&inline_headers_.name##_, Headers::get().name);                       \
   }                                                                                                \
+  void setReference##name(const std::string& value) override {                                     \
+    addSize(value.size());                                                                         \
```
Won't this double add if this is a new header per my above comment on the new addSize() call? If the header does exist, won't it not correctly account for the existing header size it is overwriting?

asraa replied:
Nope, because `maybeCreateInline(&inline_headers_.name##_, Headers::get().name)` will only account for the added key size. This call will add the value size.

asraa replied:
No! I see your point. Yes!!! Will fix, and add a testcase for setting reference and re-setting reference.

lizan commented on include/envoy/http/header_map.h:
```diff
@@ -352,6 +352,9 @@ class HeaderEntry {
   virtual HeaderEntry& insert##name() PURE;                                                        \
+  virtual void setReference##name(const std::string& value) PURE;                                  \
+  virtual void setCopy##name(absl::string_view value) PURE;                                        \
+  virtual void setInteger##name(uint64_t value) PURE;                                              \
```
Do we really need the integer interface to all headers? We only set integer value to a limited set of headers like status, grpc-status, content-length. Since now this code is generated via macro I think we can generate less?

asraa replied:
Oh interesting. Yes there's < 10 headers that need this interface, we definitely don't need all of them. I will try working this out.

asraa commented on include/envoy/http/header_map.h:
```diff
@@ -352,6 +352,9 @@ class HeaderEntry {
   virtual const HeaderEntry* name() const PURE;                                                    \
   virtual HeaderEntry* name() PURE;                                                                \
   virtual HeaderEntry& insert##name() PURE;                                                        \
+  virtual void setReference##name(const std::string& value) PURE;                                  \
```
Done. The `HeaderMapImpl::setReference` methods also take a `const std::string&` as well, since they call the `HeaderString(const std::string&)` constructor. Worth changing over?

lizan replied:
yeah, fine with just a TODO either. `const std::string&` -> `absl::string_view` is implicitly convertible but not vice versa.

jmarantz commented on source/common/grpc/common.cc:
```diff
@@ -195,21 +195,20 @@ Http::MessagePtr Common::prepareHeaders(const std::string& upstream_cluster,
   Http::MessagePtr message(new Http::RequestMessageImpl());
-  message->headers().insertMethod().value().setReference(Http::Headers::get().MethodValues.Post);
+  message->headers().setReferenceMethod(Http::Headers::get().MethodValues.Post);
   message->headers().insertPath().value().append("/", 1);
```
So my thesis here is that you should remove the insert* methods as they return hard-to-reason-about non-const refs.

In this case IMO you could just have `setPath(absl::StrCat(...))` and we're good to go :)  IDK of something was previously in 'path' though.

And if there are other cases you should expose appendPath(absl::string_view)

asraa replied:
Nothing is previously in `:path` at this point, since `message` is created right there. 
Hmmm. So there are a few small tricky `insert` cases besides this one that I wanted to handle in a separate PR (`toGrpcTimeout` in particular). I'll handle the cases that can be moved to the functionality introduced in this PR though. I think that does make sense.

jmarantz replied:
A TODO to remove the insert() methods SGTM.

asraa replied:
Done.

asraa commented:
Alright, added some markers for TODOs, updated the issue with more details and the issues https://github.com/envoyproxy/envoy/issues/8567 to track. I was using GitHub search (I do not know why I bothered..) and I didn't manage to find all the "insert"s I could fix. So I've grepped and fixed the rest of the mechanical changes that follow from the new `set##name` and `setReference##name`. PTAL

Grepping revealed that at this point:
- there are 6 non-const `insert##name` leftover (4 are appends)
- 2 non-const `get` locations.

jmarantz commented on source/common/http/utility.cc:
```diff
@@ -492,7 +492,7 @@ void Utility::transformUpgradeRequestFromH1toH2(HeaderMap& headers) {
 void Utility::transformUpgradeResponseFromH1toH2(HeaderMap& headers) {
   if (getResponseStatus(headers) == 101) {
-    headers.insertStatus().value().setInteger(200);
+    headers.setStatus(200);
```
I think maybe Lizan suggested already -- we probably don't need integer setters for every builtin header. Maybe just status and contentLength, which we could just add in explicitly as helpers.


That might (or might not) help compile/link speed :)

Also for Status we could use the http code enum rather than an int, and for both status and content-length we could maybe sanity check the input with ASSERTs.

Let's just remember that for the next round.

lizan replied:
Yeah I suggested that but it's fine to leave it for next round.

mattklein123 replied:
+1 to all those suggestions and also fine with me to defer to the next PR.

jmarantz commented:
/azp run envoy-linux
